diff --git a/configure b/configure
index 605bf3f..9842825 100755
--- a/configure
+++ b/configure
@@ -954,6 +954,7 @@ cavs_decoder_select="golomb"
 cook_decoder_select="fft mdct"
 cscd_decoder_suggest="zlib"
 dca_decoder_select="fft mdct"
+dirac_decoder_select="golomb"
 dnxhd_encoder_select="aandct"
 dxa_decoder_select="zlib"
 eac3_decoder_deps="gpl"
diff --git a/libavcodec/Makefile b/libavcodec/Makefile
index b7cce5f..a001cba 100644
--- a/libavcodec/Makefile
+++ b/libavcodec/Makefile
@@ -58,6 +58,7 @@ OBJS-$(CONFIG_COOK_DECODER)            += cook.o
 OBJS-$(CONFIG_CSCD_DECODER)            += cscd.o
 OBJS-$(CONFIG_CYUV_DECODER)            += cyuv.o
 OBJS-$(CONFIG_DCA_DECODER)             += dca.o
+OBJS-$(CONFIG_DIRAC_DECODER)           += diracdec.o dirac.o dirac_arith.o dirac_wavelet.o mpeg12data.o
 OBJS-$(CONFIG_DNXHD_DECODER)           += dnxhddec.o dnxhddata.o
 OBJS-$(CONFIG_DNXHD_ENCODER)           += dnxhdenc.o dnxhddata.o mpegvideo_enc.o motion_est.o ratecontrol.o mpeg12data.o mpegvideo.o
 OBJS-$(CONFIG_DSICINAUDIO_DECODER)     += dsicinav.o
diff --git a/libavcodec/allcodecs.c b/libavcodec/allcodecs.c
index d5fe8b5..fc8a12a 100644
--- a/libavcodec/allcodecs.c
+++ b/libavcodec/allcodecs.c
@@ -74,6 +74,7 @@ void avcodec_register_all(void)
     REGISTER_DECODER (CLJR, cljr);
     REGISTER_DECODER (CSCD, cscd);
     REGISTER_DECODER (CYUV, cyuv);
+    REGISTER_DECODER (DIRAC, dirac);
     REGISTER_ENCDEC  (DNXHD, dnxhd);
     REGISTER_DECODER (DSICINVIDEO, dsicinvideo);
     REGISTER_ENCDEC  (DVVIDEO, dvvideo);
