Index: libavcodec/golomb.h
===================================================================
--- libavcodec/golomb.h	(revision 11312)
+++ libavcodec/golomb.h	(working copy)
@@ -504,4 +504,31 @@
     set_ur_golomb_jpegls(pb, v, k, limit, esc_len);
 }
 
+/**
+ * write unsigned interleaved exponential golomb code (Dirac)
+ */
+static inline void dirac_set_ue_golomb(PutBitContext *pb, int i)
+{
+    int log = av_log2(++i);
+    while(log) {
+        log--;
+        put_bits(pb, 1, 0);
+        put_bits(pb, 1, (i >> log)&1);
+    }
+    put_bits(pb, 1, 1);
+}
+
+/**
+ * write signed interleaved exponential golomb code (Dirac)
+ */
+static inline void dirac_set_se_golomb(PutBitContext *pb, int i)
+{
+    dirac_set_ue_golomb(pb, FFABS(i));
+    if (i > 0)
+        put_bits(pb, 1, 0);
+    else if (i < 0)
+        put_bits(pb, 1, 1);
+}
+
+
 #endif /* FFMPEG_GOLOMB_H */
Index: libavcodec/Makefile
===================================================================
--- libavcodec/Makefile	(revision 11312)
+++ libavcodec/Makefile	(working copy)
@@ -56,6 +56,8 @@
 OBJS-$(CONFIG_CSCD_DECODER)            += cscd.o
 OBJS-$(CONFIG_CYUV_DECODER)            += cyuv.o
 OBJS-$(CONFIG_DCA_DECODER)             += dca.o
+OBJS-$(CONFIG_DIRAC_DECODER)           += diracdec.o dirac.o dirac_arith.o dirac_wavelet.o golomb.o
+OBJS-$(CONFIG_DIRAC_ENCODER)           += diracenc.o dirac.o dirac_arith.o dirac_wavelet.o golomb.o
 OBJS-$(CONFIG_DNXHD_DECODER)           += dnxhddec.o dnxhddata.o
 OBJS-$(CONFIG_DNXHD_ENCODER)           += dnxhdenc.o dnxhddata.o mpegvideo_enc.o motion_est.o ratecontrol.o
 OBJS-$(CONFIG_DSICINAUDIO_DECODER)     += dsicinav.o
@@ -314,6 +316,7 @@
 OBJS-$(CONFIG_AC3_PARSER)              += ac3_parser.o ac3tab.o aac_ac3_parser.o
 OBJS-$(CONFIG_CAVSVIDEO_PARSER)        += cavs_parser.o
 OBJS-$(CONFIG_DCA_PARSER)              += dca_parser.o
+OBJS-$(CONFIG_DIRAC_PARSER)            += dirac_parser.o
 OBJS-$(CONFIG_DVBSUB_PARSER)           += dvbsub_parser.o
 OBJS-$(CONFIG_DVDSUB_PARSER)           += dvdsub_parser.o
 OBJS-$(CONFIG_H261_PARSER)             += h261_parser.o
Index: libavcodec/allcodecs.c
===================================================================
--- libavcodec/allcodecs.c	(revision 11312)
+++ libavcodec/allcodecs.c	(working copy)
@@ -73,6 +73,7 @@
     REGISTER_DECODER (CLJR, cljr);
     REGISTER_DECODER (CSCD, cscd);
     REGISTER_DECODER (CYUV, cyuv);
+    REGISTER_ENCDEC  (DIRAC, dirac);
     REGISTER_ENCDEC  (DNXHD, dnxhd);
     REGISTER_DECODER (DSICINVIDEO, dsicinvideo);
     REGISTER_ENCDEC  (DVVIDEO, dvvideo);
@@ -283,6 +284,7 @@
     REGISTER_PARSER  (AC3, ac3);
     REGISTER_PARSER  (CAVSVIDEO, cavsvideo);
     REGISTER_PARSER  (DCA, dca);
+    REGISTER_PARSER  (DIRAC, dirac);
     REGISTER_PARSER  (DVBSUB, dvbsub);
     REGISTER_PARSER  (DVDSUB, dvdsub);
     REGISTER_PARSER  (H261, h261);
Index: libavcodec/avcodec.h
===================================================================
--- libavcodec/avcodec.h	(revision 11312)
+++ libavcodec/avcodec.h	(working copy)
@@ -169,6 +169,7 @@
     CODEC_ID_VP6A,
     CODEC_ID_AMV,
     CODEC_ID_VB,
+    CODEC_ID_DIRAC,
 
     /* various PCM "codecs" */
     CODEC_ID_PCM_S16LE= 0x10000,
Index: tests/regression.sh
===================================================================
--- tests/regression.sh	(revision 11312)
+++ tests/regression.sh	(working copy)
@@ -40,6 +40,8 @@
     do_snow=y
 elif [ "$1" = "snowll" ] ; then
     do_snowll=y
+elif [ "$1" = "dirac" ] ; then
+    do_dirac=y
 elif [ "$1" = "libavtest" ] ; then
     do_libavtest=y
     logfile="$datadir/libav.regression"
@@ -85,6 +87,7 @@
     do_snow=y
     do_snowll=y
     do_adpcm_yam=y
+    do_dirac=y
     do_dv=y
     do_dv50=y
     do_flashsv=y
@@ -493,6 +496,13 @@
 fi
 
 ###################################
+if [ -n "$do_dirac" ] ; then
+# dirac
+do_video_encoding dirac.drc "-strict -2" pgmyuv "-an -vcodec dirac -s 128x64 -f rawvideo"
+do_video_decoding "" "-s 352x288"
+fi
+
+###################################
 if [ -n "$do_dv" ] ; then
 # dv
 do_video_encoding dv.dv "-dct int" pgmyuv "-s pal -an"
Index: libavformat/riff.c
===================================================================
--- libavformat/riff.c	(revision 11312)
+++ libavformat/riff.c	(working copy)
@@ -37,6 +37,7 @@
     { CODEC_ID_H263P, MKTAG('H', '2', '6', '3') },
     { CODEC_ID_H263I, MKTAG('I', '2', '6', '3') }, /* intel h263 */
     { CODEC_ID_H261, MKTAG('H', '2', '6', '1') },
+    { CODEC_ID_DIRAC, MKTAG('D', 'R', 'A', 'C') },
 
     /* added based on MPlayer */
     { CODEC_ID_H263P, MKTAG('U', '2', '6', '3') },
Index: libavformat/raw.c
===================================================================
--- libavformat/raw.c	(revision 11312)
+++ libavformat/raw.c	(working copy)
@@ -312,6 +312,7 @@
         av_set_pts_info(st, 64, ap->time_base.num, ap->time_base.den);
     } else if ( st->codec->codec_id == CODEC_ID_MJPEG ||
                 st->codec->codec_id == CODEC_ID_MPEG4 ||
+                st->codec->codec_id == CODEC_ID_DIRAC ||
                 st->codec->codec_id == CODEC_ID_H264) {
         av_set_pts_info(st, 64, 1, 25);
     }
@@ -404,6 +405,14 @@
     return 0;
 }
 
+static int dirac_probe(AVProbeData *p)
+{
+    if (AV_RL32(p->buf) == MKTAG('B', 'B', 'C', 'D'))
+        return AVPROBE_SCORE_MAX;
+    else
+        return 0;
+}
+
 static int ac3_probe(AVProbeData *p)
 {
     int max_frames, first_frames = 0, frames;
@@ -506,6 +515,33 @@
 };
 #endif //CONFIG_MUXERS
 
+AVInputFormat dirac_demuxer = {
+    "dirac",
+    "raw dirac",
+    0,
+    dirac_probe,
+    video_read_header,
+    raw_read_partial_packet,
+    raw_read_close,
+    .flags= AVFMT_GENERIC_INDEX,
+    .value = CODEC_ID_DIRAC,
+};
+
+#ifdef CONFIG_MUXERS
+AVOutputFormat dirac_muxer = {
+    "dirac",
+    "raw dirac",
+    NULL,
+    "drc",
+    0,
+    0,
+    CODEC_ID_DIRAC,
+    NULL,
+    raw_write_packet,
+    .flags= AVFMT_NOTIMESTAMPS,
+};
+#endif
+
 AVInputFormat dts_demuxer = {
     "dts",
     "raw dts",
Index: libavformat/allformats.c
===================================================================
--- libavformat/allformats.c	(revision 11312)
+++ libavformat/allformats.c	(working copy)
@@ -67,6 +67,7 @@
     REGISTER_DEMUXER  (C93, c93);
     REGISTER_MUXER    (CRC, crc);
     REGISTER_DEMUXER  (DAUD, daud);
+    REGISTER_MUXDEMUX (DIRAC, dirac);
     REGISTER_DEMUXER  (DSICIN, dsicin);
     REGISTER_DEMUXER  (DTS, dts);
     REGISTER_MUXDEMUX (DV, dv);
Index: Makefile
===================================================================
--- Makefile	(revision 11312)
+++ Makefile	(working copy)
@@ -285,7 +285,7 @@
 	@echo
 	$(SRC_PATH)/tests/server-regression.sh $(FFSERVER_REFFILE) $(SRC_PATH)/tests/test.conf
 
-codectest mpeg4 mpeg ac3 snow snowll: ffmpeg$(EXESUF) tests/vsynth1/00.pgm tests/vsynth2/00.pgm tests/asynth1.sw tests/tiny_psnr$(EXESUF)
+codectest mpeg4 mpeg ac3 snow snowll dirac: ffmpeg$(EXESUF) tests/vsynth1/00.pgm tests/vsynth2/00.pgm tests/asynth1.sw tests/tiny_psnr$(EXESUF)
 	$(SRC_PATH)/tests/regression.sh $@ $(FFMPEG_REFFILE)   tests/vsynth1
 	$(SRC_PATH)/tests/regression.sh $@ $(ROTOZOOM_REFFILE) tests/vsynth2
 
@@ -296,7 +296,7 @@
 	$(SRC_PATH)/tests/seek_test.sh $(SEEK_REFFILE)
 
 ifeq ($(CONFIG_SWSCALER),yes)
-test-server codectest mpeg4 mpeg ac3 snow snowll libavtest: swscale_error
+test-server codectest mpeg4 mpeg ac3 snow snowll dirac libavtest: swscale_error
 swscale_error:
 	@echo
 	@echo "This regression test is incompatible with --enable-swscaler."
@@ -325,7 +325,7 @@
 .PHONY: all lib videohook documentation install* wininstaller uninstall*
 .PHONY: dep depend clean distclean TAGS
 .PHONY: codectest libavtest seektest test-server fulltest test
-.PHONY: mpeg4 mpeg ac3 snow snowll swscale-error
+.PHONY: mpeg4 mpeg ac3 snow snowll dirac swscale-error
 
 -include .depend
 -include .vhookdep
